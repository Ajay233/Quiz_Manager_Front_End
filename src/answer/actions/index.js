import { getUsingParams, post, put, del } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import { hideModal } from '../../modal/actions'
import { sessionExpired } from '../../utils/session'
import history from '../../history'

// create
export const addAnswer = (data, jwt) => {
  return (dispatch) => {
    return post("answer/create", [data], jwt).then((response) => {
      dispatch({
        type: "ADD_ANSWER",
        payload: response.data[0]
      })
      history.push("/editQuestion");
      dispatch(setNotification("Answer created", "success", true));
    }).catch((error) => {
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        console.log(error.response);
        dispatch(setNotification("Error - unable to create answer", "error", true));
      }
    });
  }
}

// read
export const getAnswers = (param, jwt) => {
  return (dispatch) => {
    return getUsingParams("answer/findByQuestionId", param, jwt).then((response) => {
      dispatch({
        type: "SET_ANSWERS",
        payload: response.data
      });
    }).catch((error) => {
      dispatch(setNotification("Error retrieving answers", "error", true));
    });
  }
}

// update (there was a call to setCurrentAnswer but have removed this as it's not needed)
export const updateAnswer = (body, jwt) => {
  return (dispatch) => {
    return put("answer/update", [body], jwt).then((response) => {
      dispatch({
        type: "UPDATE_ANSWER",
        payload: response.data[0]
      })
      history.push("/editQuestion");
      dispatch(setNotification("Answer updated", "success", true));
    }).catch((error) => {
      console.log(error.response);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(setNotification("Error - unable to update answer", "error", true));
      }
    });
  }
}

// delete
export const deleteAnswer = (config, jwt) => {
  return (dispatch) => {
    return del("answer/delete", config, jwt).then((response) => {
      dispatch(hideModal())
      dispatch({
        type: "DELETE_ANSWER",
        payload: config.data[0]
      })
      dispatch(setNotification("Answer deleted", "success", true));
    }).catch((error) => {
      console.log(error.response);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(hideModal())
        dispatch(setNotification("Error - Unable to delete this answer", "error", true))
      }
    });
  }
}

export const setCurrentAnswer = (answer) => {
  return {
    type: "SET_CURRENT_ANSWER",
    payload: answer
  }
}
