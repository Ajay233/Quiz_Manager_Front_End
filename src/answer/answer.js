import React from 'react'
import { Link } from 'react-router-dom'

export class Answer extends React.Component {

  handleEdit = () => {
    this.props.setCurrentAnswer(this.props.answer);
  }

  handleDelete = () => {
    this.props.showModal();
    this.props.setCurrentAnswer(this.props.answer);
  }

  renderAnswer = () => {
    const { answerNumber, description, correctAnswer } = this.props.answer
    return(
      <div className="answerContainer">
        <div className="ansNumber">{ answerNumber }</div>
        <div className="ansDecription">{ description }</div>
        <div className="correctAns">{ correctAnswer === true ? <i className="far fa-check-circle green"></i> : <i className="far fa-times-circle red"></i> }</div>
        {this.renderOptions()}
      </div>
    );
  }

  renderOptions = () => {
    const { permission } = this.props.userData;
    return(
      <div className="options">
        { permission === "ADMIN" || permission === "SUPER-USER" ? <Link to="#" className="deleteOption" onClick={this.handleDelete}><i className="fas fa-trash-alt red"></i> Delete</Link> : null }
        { permission === "ADMIN" || permission === "SUPER-USER" ? <Link to="/editAnswer" className="edit" onClick={this.handleEdit}><i className="fas fa-edit blue"></i> Edit</Link> : null }
      </div>
    );
  }

  render(){
    return(
      <div>
        {this.renderAnswer()}
      </div>
    );
  }
}

export default Answer
