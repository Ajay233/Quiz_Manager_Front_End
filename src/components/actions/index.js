export const setToTopButton = (show=false) => {
  return {
    type: "SHOW_TO_TOP_BUTTON",
    payload: show
  }
}

export const setUrl = (url="") => {
  return {
    type: "SET_URL",
    payload: url
  }
}

export const enableLightMode = () => {
  return {
    type: "SET_THEME",
    payload: false
  }
}

export const enableDarkMode = () => {
  return {
    type: "SET_THEME",
    payload: true
  }
}

export const setLoaderState = (show=false, message="", label="") => {
  return {
    type: "SET_LOADER_STATE",
    payload: {
      show: show,
      message: message,
      label: label
    }
  }
}
