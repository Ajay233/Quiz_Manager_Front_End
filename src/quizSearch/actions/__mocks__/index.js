export const setQuizes = jest.fn()

export const getAllQuizes = jest.fn()

export const getQuizSearchResults = jest.fn()

export const deleteQuiz = jest.fn()

export const addQuiz = jest.fn()

export const clearQuizes = jest.fn()
