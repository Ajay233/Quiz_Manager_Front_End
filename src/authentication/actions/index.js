import { post } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import history from '../../history'

export const setUser = (endpoint, loginDetails) => {

  return (dispatch) => {

    post(endpoint, loginDetails).then((response) => {
      dispatch({
        type: "SET_USER_LOGGED_IN",
        payload: {
          id: response.data.user.id,
          forename: response.data.user.forename,
          surname: response.data.user.surname,
          email: response.data.user.email,
          permission: response.data.user.permission,
          verified: response.data.user.verified,
          jwt: response.data.jwt,
          loggedIn: true,
          response: response
        }
      })
      history.push("/")
      dispatch(setNotification(`Welcome back ${response.data.user.forename}`, "loginSuccess", true))
    }).catch((error) => {
      if(error.response.data === "NOT VERIFIED"){
        const verifyMsg = "Your email has not yet been veirified.  You will need to verify your email before you can log in"
        dispatch(setNotification(verifyMsg, "warning", true))
      } else {
        dispatch({
          type: "SET_USER_LOGGED_IN",
          payload: {
            error: error.response
          }
        });
        dispatch(setNotification("The username or password you entered was incorrect", "error", true));
      }
    })
  }
}

export const logOutUser = () => {
  return {
    type: "LOG_OUT_USER",
    payload: {
      forename: "",
      surname: "",
      email: "",
      permission: "",
      verified: "",
      jwt: "",
      loggedIn: false
    }
  }
}

export const setVerficationProcess = (endpoint, verificationDetails) => {

  return (dispatch) => {

    post(endpoint, verificationDetails).then((response) => {
      dispatch({
        type: "SET_VERIFY_PROCESS_STATUS",
        payload: {
          completionStatus: "completed",
          token: "",
          error: {}
        }
      })
    }).catch((error) => {
      dispatch({
        type: "SET_VERIFY_PROCESS_STATUS",
        payload: {
          completionStatus: "not verified",
          token: verificationDetails.token,
          error: error.response
        }
      })
    })
  }
}

export const setRedirect = (status) => {
  return {
    type: "SET_REDIRECT",
    payload: {
      status: status
    }
  }
}

export const logOut = () => {
  return {
    type: "RESET_APP"
  }
}

// export const getUsers = (endpoint, header) => {
//   return {
//     type: "GET_USER",
//     payload: {
//       userId
//   }
// }
