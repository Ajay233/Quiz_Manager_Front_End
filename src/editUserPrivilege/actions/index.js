import { getUsingParams } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import { sessionExpired } from '../../utils/session'

export const setUserResults = (param, jwt) => {
  return (dispatch) => {
    return getUsingParams("users/findByEmail", param, jwt).then((response) => {
      dispatch({
        type: "SET_USER_RESULTS",
        payload: response.data
      })
    }).catch((error) => {
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        const msg = error.response.data;
        dispatch(setNotification(msg, "error", true));
        console.log(error.response)
      }
    });
  }
}

export const clearUserResults = () => {
  return {
    type: "CLEAR_USER_RESULTS",
    payload: {}
  }
}
