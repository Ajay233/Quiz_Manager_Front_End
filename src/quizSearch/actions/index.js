import { get, getUsingParams } from '../../axiosRequests/requests';
import { setNotification } from '../../notifications/actions'
import { sessionExpired } from '../../utils/session'

export const setQuizes = (quizes) => {
  return {
    type: "SET_QUIZES",
    payload: quizes
  }
}

export const getAllQuizes = (endpoint, token) => {
  return (dispatch) => {
    return get(endpoint, token).then((response) => {
      dispatch({
        type: "GET_ALL_QUIZES",
        payload: response.data
      })
    }).catch((error) => {
      console.log(error.response)
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(setNotification("Error - unable to get quiz list", "error", true));
      }

    });
  }
}

export const getQuizSearchResults = (endpoint, data, token, permission, errorMsg) => {
  return (dispatch) => {
    return getUsingParams(endpoint, data, token).then((response) => {
      let filteredQuizList = permission === "USER" ? response.data.filter(quiz => quiz.status === "READY") : response.data;
      dispatch({
        type: "SET_QUIZ_SEARCH_RESULTS",
        payload: filteredQuizList
      })
      if(filteredQuizList.length === 0){
         dispatch(setNotification(errorMsg, "error", true))
       }
    }).catch((error) => {
      console.log(error.response);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
      dispatch(setNotification(errorMsg, "error", true));
      }
    })
  }
}

export const deleteQuiz = (quiz) => {
  return {
    type: "DELETE_QUIZ",
    payload: quiz
  }
}

export const addQuiz = (quiz) => {
  return {
    type: "ADD_QUIZ",
    payload: quiz
  }
}

export const clearQuizes = () => {
  return {
    type: "CLEAR_QUIZZES_STORE",
    payload: []
  }
}

export const deleteQuizFromCategory = (quiz) => {
  return {
    type: "DELETE_QUIZ_FROM_CATEGORY",
    payload: quiz
  }
}
