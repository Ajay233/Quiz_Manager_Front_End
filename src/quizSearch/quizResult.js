import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import Modal from '../modal/modal'
import { setQuiz } from '../quiz/actions'
import { del } from '../axiosRequests/requests'
import { setNotification } from '../notifications/actions'
import { deleteQuiz } from './actions'
import { showModal } from '../modal/actions'

class QuizResult extends React.Component {

  handleEdit = () => {

  }

  renderModal = () => {

  }

  // handleDelete = () => {
  //   const { quiz, userData, deleteQuiz, setNotification } = this.props;
  //   const config = {
  //     data: quiz
  //   }
  //   del("quiz/delete", config, userData.jwt).then((response) => {
  //     deleteQuiz(quiz)
  //     setNotification("Quiz deleted", "success", true)
  //   }).catch((error) => {
  //     console.log(error.response)
  //     setNotification("Error - Unable to delete this quiz", "error", true)
  //   })
  // }

  handleDelete = () => {
    this.props.setQuiz(this.props.quiz);
    this.props.showModal();
  }

  handleView = () => {
    this.props.setQuiz(this.props.quiz);
  }

  handleStart = () => {
    console.log(this.props)
  }

  renderQuiz = () => {
    const { name, description} = this.props.quiz
    return(
      <div className="quiz">
        <div className="quizName">{name}</div>
        <div className="quizDescription">{description}</div>
        {this.renderOptions()}
      </div>
    );
  }

  // possible way to use different classes for option sheader if the header ends up being used
  optionsClass = (permission) => {
    return permission === "ADMIN" ? "" : "";
  }

  renderOptions = () => {
    const { permission } = this.props;
    return(
      <div className="options">
        { permission === "USER" ? <Link to="#" className="start" onClick={this.handleStart}><i className="far fa-play-circle blue"></i> Start</Link> : null }
        { permission === "READ-ONLY" ? <Link to="/viewQuiz" className="view" onClick={this.handleView}><i className="far fa-eye blue"></i> View</Link> : null }
        { permission === "ADMIN" ? <Link to="/editQuiz" className="edit" onClick={this.handleEdit}><i className="fas fa-edit blue"></i> Edit</Link> : null }
        { permission === "ADMIN" ? <Link to="#" className="deleteOption" onClick={this.handleDelete}><i className="fas fa-trash-alt red"></i> Delete</Link> : null }
      </div>
    );
  }

  render(){
    return(
      <div className="quizContainer">
        {this.renderQuiz()}
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    userData: state.userData,
    modalState: state.showModal
  }
}

export default connect(mapStateToProps,{ setQuiz, setNotification, deleteQuiz, showModal })(QuizResult)
