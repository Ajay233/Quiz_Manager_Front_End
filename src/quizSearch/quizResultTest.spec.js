import React from 'react'
import ReactDOM from 'react-dom'
import { QuizResult, mapStateToProps } from './quizResult'
import { Router } from 'react-router-dom'
import { Provider } from 'react-redux'
import { mount } from 'enzyme'
import configureStore from 'redux-mock-store'
import mockAxios from 'jest-mock-axios'
import { render, fireEvent, cleanup } from '@testing-library/react'
import renderer from 'react-test-renderer'
import history from '../history'

import { setQuiz } from '../quiz/actions'
import { setNotification } from '../notifications/actions'
import { deleteQuiz } from './actions'
import { showModal, showModal2 } from '../modal/actions'
import { getQuestions, setCurrentQuestion } from '../question/actions'
import { getAnswers } from '../answer/actions'

jest.mock("../quiz/actions")
jest.mock("../notifications/actions")
jest.mock("./actions")
jest.mock("../modal/actions")
jest.mock("../question/actions")
jest.mock("../answer/actions")

const mockStore = configureStore({})

describe("mapStateToProps", () => {
  it("should mapStateToProps", () => {
    const appState = {
      userData: { id: 1, permission: "USER" },
      modalState: { showModal: false, showModal2: false, showModal3: false}
    }

    const componentState = mapStateToProps(appState)
    expect(componentState).toEqual(appState)
  })
})

describe("QuizResult", () => {

  let store;
  let quiz;
  beforeEach(() => {
    quiz = { id: 1, name: "test", description: "test", category: "test", status: "READY" }
  })

  afterEach(() => {
    showModal2.mockClear()
    showModal.mockClear()
    setQuiz.mockClear()
  })

  it("should render a quiz result and the appropriate options for a user with USER permission", () => {

    store = mockStore({
      userData: { id: 1, permission: "USER", jwt: "jwt" },
      modalState: { showModal: false, showModal2: false, showModal3: false}
    })

    const component = renderer.create(
      <Provider store={store}>
        <Router history={history}>
          <QuizResult
            quiz={quiz}
            permission={"USER"}
          />
        </Router>
      </Provider>
    )

    expect(component).toMatchSnapshot()
  })

  it("should render a quiz result and the appropriate options for a user with READ-ONLY permission", () => {
    store = mockStore({
      userData: { id: 1, permission: "READ-ONLY" },
      modalState: { showModal: false, showModal2: false, showModal3: false}
    })

    const component = renderer.create(
      <Provider store={store}>
        <Router history={history}>
          <QuizResult
            quiz={quiz}
            permission={"READ-ONLY"}
          />
        </Router>
      </Provider>
    )

    expect(component).toMatchSnapshot()
  })

  it("should render a quiz result and the appropriate options for a user with ADMIN permission", () => {
    store = mockStore({
      userData: { id: 1, permission: "ADMIN" },
      modalState: { showModal: false, showModal2: false, showModal3: false}
    })

    const component = renderer.create(
      <Provider store={store}>
        <Router history={history}>
          <QuizResult
            quiz={quiz}
            permission={"ADMIN"}
          />
        </Router>
      </Provider>
    )

    expect(component).toMatchSnapshot()
  })

  describe("handleStart", () => {

    it("should call", () => {
      const component = render(
        <Provider store={store}>
          <Router history={history}>
            <QuizResult
              quiz={quiz}
              permission={"USER"}
              setQuiz={setQuiz}
              showModal2={showModal2}
            />
          </Router>
        </Provider>
      )

      fireEvent.click(component.getByText("Start"))
      expect(setQuiz).toHaveBeenCalledTimes(1)
      expect(setQuiz).toHaveBeenCalledWith(quiz)
      expect(showModal2).toHaveBeenCalledTimes(1)
    })
  })

  describe("handleView", () => {
    it("should call", () => {

      const userDataProp = { id: 1, permission: "USER", jwt: "jwt" }

      const component = render(
        <Provider store={store}>
          <Router history={history}>
            <QuizResult
              quiz={quiz}
              permission={"READ-ONLY"}
              setQuiz={setQuiz}
              getQuestions={getQuestions}
              userData={userDataProp}
              setNotification={setNotification}
            />
          </Router>
        </Provider>
      )

      const param = { quizId: 1 }

      fireEvent.click(component.getByText("View"))
      expect(getQuestions).toHaveBeenCalledTimes(1)
      expect(getQuestions).toHaveBeenCalledWith("question/findByQuizId", param, "jwt")
      expect(setQuiz).toHaveBeenCalledTimes(1)
      expect(setQuiz).toHaveBeenCalledWith(quiz)
    })
  })

  describe("handleDelete", () => {
    it("should call", () => {
      const component = render(
        <Provider store={store}>
          <Router history={history}>
            <QuizResult
              quiz={quiz}
              permission={"ADMIN"}
              setQuiz={setQuiz}
              showModal={showModal}
            />
          </Router>
        </Provider>
      )

      fireEvent.click(component.getByText("Delete"))
      expect(setQuiz).toHaveBeenCalledTimes(1)
      expect(setQuiz).toHaveBeenCalledWith(quiz)
      expect(showModal).toHaveBeenCalledTimes(1)
    })
  })

})
