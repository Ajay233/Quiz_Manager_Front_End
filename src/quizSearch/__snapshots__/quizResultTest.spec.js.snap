// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuizResult should render a quiz result and the appropriate options for a user with ADMIN permission 1`] = `
<div
  className="quizContainer"
>
  <div
    className="quiz"
  >
    <div
      className="quizName"
    >
      test
    </div>
    <div
      className="quizDescription"
    >
      test
    </div>
    <div
      className="options"
    >
      <a
        className="deleteOption"
        href="/"
        onClick={[Function]}
      >
        <i
          className="fas fa-trash-alt red"
        />
         Delete
      </a>
      <a
        className="edit"
        href="/editQuiz"
        onClick={[Function]}
      >
        <i
          className="fas fa-edit blue"
        />
         Edit
      </a>
      <a
        className="start"
        href="/"
        onClick={[Function]}
      >
        <i
          className="far fa-play-circle blue"
        />
         Start
      </a>
    </div>
  </div>
</div>
`;

exports[`QuizResult should render a quiz result and the appropriate options for a user with READ-ONLY permission 1`] = `
<div
  className="quizContainer"
>
  <div
    className="quiz"
  >
    <div
      className="quizName"
    >
      test
    </div>
    <div
      className="quizDescription"
    >
      test
    </div>
    <div
      className="options"
    >
      <a
        className="view"
        href="/viewQuiz"
        onClick={[Function]}
      >
        <i
          className="far fa-eye blue"
        />
         View
      </a>
      <a
        className="start"
        href="/"
        onClick={[Function]}
      >
        <i
          className="far fa-play-circle blue"
        />
         Start
      </a>
    </div>
  </div>
</div>
`;

exports[`QuizResult should render a quiz result and the appropriate options for a user with USER permission 1`] = `
<div
  className="quizContainer"
>
  <div
    className="quiz"
  >
    <div
      className="quizName"
    >
      test
    </div>
    <div
      className="quizDescription"
    >
      test
    </div>
    <div
      className="options"
    >
      <a
        className="start"
        href="/"
        onClick={[Function]}
      >
        <i
          className="far fa-play-circle blue"
        />
         Start
      </a>
    </div>
  </div>
</div>
`;
